### PR TITLE
coord: Split coord.rs into multiple modules

### PR DESCRIPTION
**Pull request by jkosh44**
### Description
The refactor is based on the following structure:
coord rs refactor

See commit message

### Motivation
* This PR refactors existing code.

### Tips for reviewer

* No code has been changed, only moved to a different file or extracted into a method. You can try looking through all the diff, but that will probably be painful. Instead looking at file names and overall structure would be more useful.

### Testing

- [X] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Comment by jkosh44** [comment]
@mjibson @benesch This is a POC I came up with for a potential coord.rs refactor in case you're curious.

**Review by benesch — APPROVED** [review]
LGTM but I didn't review in detail. I'm in favor of doing _something_ to cut down on the size of coord.rs—and this seems quite sensible—and then doing smaller more targeted refactors once it's a more manageable size.

As a general philosophy, I don't love is splitting the `impl` of a type across multiple modules. For me the cognitive overhead is not the length of coord.rs, but the fact that understanding `struct Coordinator` requires understanding the dozens and dozens and dozens of methods all together. I know opinions here vary, but I find that splitting `impl`s across modules just sweeps that complexity under the rug.

I'm still very much in favor of landing this, especially as the `Coordinator`'s `impl` blocks are already split across several modules—just want us to continue to keep an eye out for chunks of the coordinator that can be split out into their own sub-structs with smaller, better-documented APIs.

**Review by frankmcsherry — APPROVED** [review]
I looked at the structure, and aside from @benesch' concern that some implementations are a bit more spilled across a few files, I think it is a good thing to do. Many of the broken-out files are great to break out, and .. we can iterate on how to localize the reasoning about what `Coordinator` actually needs of each of its parts (I'd love to extract some "high priority core", where we e.g. assign timestamps, transact with the controller, and .. at some level aren't too bothered about what caused us to choose the specific actions we performed). For the moment, this seems like a great first step, though!

**Comment by jkosh44** [comment]
Cool, I think I'm going to redo this over the weekend or next Friday for Skunkworks. It's almost impossible to keep the PR up to date and I don't want it to be too disruptive during the week.

**Comment by jkosh44** [comment]
@frankmcsherry @benesch @mjibson 

I've made the following changes to the above diagram:
- Combined `serve` and `bootstrap` in `coord.rs`.
- Attempted to split apart the `util` module.

I don't expect you to look at every single line but was hoping I could get some feedback on:
- The overall structure of the refactor.
- The doc comments at the top of each module.
- Should I keep all the separate utility modules or just combine them into one large module like in the diagram? There is a bit of a gross inter-dependency between the utility modules that I'm worried may make them a pain to use.


Below, for reference, is structure of the Coordinator that is implemented by this
commit. For each module I've included a brief description (mostly
copied from doc comments), the structs contained in the modules, and
the Coordinator methods contained in the module.

# Root Module
This is the root module for the Coordinator and resides in coord.rs

## coord

### Description
The root of the Coordinator.

### Structs
- Message
- SendDiffs
- CreateSourceStatementReady
- SinkConnectionReady
- Config
- ReplicaMetadata
- ConnMeta
- TxnReads
- Coordinator

### Methods
- bootstrap
- serve
- create_storage_export

# Message Flow Modules
These modules contain the logic for how a message is received,
processed, and executed by the Coordinator.

## message_handler

### Description
Logic for processing `Coordinator` messages. The `Coordinator`
receives messages from various sources (ex: controller, clients,
background tasks, etc).

### Structs
None

### Methods
- handle_message
- message_command
- message_controller
- message_create_source_statement_ready
- message_sink_connection_ready
- message_write_lock_grant
- message_send_diffs
- message_advance_timelines
- message_compute_instance_status


## command_handler

### Description
Logic for  processing client `Command`s. Each `Command` is
initiated by a client via some external Materialize API (ex: HTTP and
psql).

### Structs
None

### Methods
- handle_command
- handle_startup
- handle_execute
- handle_execute_inner
- handle_describe
- handle_cancel
- handle_terminate

## sequencer

### Description
Logic for executing a planned SQL query.

### Struct
None

### Methods
- sequence_plan
- sequence_<X> where X is the name of some plan. I've left all the
specific methods out because there's so many of them.

# Utility Modules
These modules contain the logic and APIs for various
responsibilities and functionalities taken on by the Coordinator.
Unfortunately there is a gross interdependency between many of these
modules. I believe with some work we can tease out many of these
interdependencies. It's possible that the correct design is to combine
these all into a single large module.

## appends

### Description
Logic and types for all appends executed by the `Coordinator`.

### Structs
- AdvanceLocalInput
- AdvanceTables
- Deferred
- DeferredPlan
- PendingWriteTxn

### Macros
- guard_write_critical_section

### Methods
- try_group_commit
- group_commit
- submit_write
- send_builtin_table_updates
- queue_local_input_advances
- advance_local_inputs
- defer_write
- try_grant_session_write_lock

## dataflow (formerly dataflow_builder)

### Description
Types and methods for building and shipping dataflow descriptions.

Dataflows are buildable from the coordinator's `catalog` and `indexes`
members, which respectively describe the collection backing identifiers
and indicate which identifiers have arrangements available. This module
isolates that logic from the rest of the somewhat complicated
coordinator.

### Structs
- DataflowBuilder
- ExprPrepStyle

### Methods
- dataflow_builder
- ship_dataflow
- ship_dataflows
- finalize_dataflow

### Functions
- prep_relation_expr
- prep_scalar_expr
- eval_unmaterializable_func

## ddl

### Description
This module encapsulates all of the `Coordinator`'s logic for
creating, dropping, and altering objects.

### Structs
- CatalogTxn

### Methods
- catalog_transact
- drop_sources
- drop_sinks
- drop_indexes
- drop_materialized_views
- drop_secrets
- drop_temp_items
- handle_sink_connection_ready

## id_bundle

### Description
A bundle of storage and compute collection identifiers (this was a
pre-existing module).

### Structs
CollectionIdBundle

### Methods
None

## indexes

### Description
Logic related to indexes (this was a pre-existing module).

### Structs
- ComputeInstanceIndexOracle

### Methods
- index_oracle

## peek (formerly fast_path_peek)

### Description
Logic and types for creating, executing, and tracking peeks.

This module determines if a dataflow can be short-cut, by returning
constant values or by reading out of existing arrangements, and
implements the appropriate plan.

### Structs
- PendingPeek
- PeekResponseUnary
- PeekDataflowPlan
- Plan

### Methods
- implement_fast_path_peek
- remove_pending_peeks

### Functions
- create_plan

## read_policy

### Description
Types and methods related to initializing, updating, and removing read
policies on collections.

This module contains the API for read holds on collections. A "read
hold" prevents the controller from compacting the associated
collections, and ensures that they remain "readable" at a specific
time, as long as the hold is held.

### Structs
- ReadCapability
- ReadHolds

### Methods
- initialize_storage_read_policies
- initialize_compute_read_policies
- initialize_read_policies
- acquire_read_holds
- update_read_holds
- release_read_holds

## timeline

### Description
A mechanism to ensure that a sequence of writes and reads proceed
correctly through timestamps.

### Structs
- WriteTimestamp
- TimelineState
- TimestampOracleState
- TimestampOracle
- DurableTimestampOracle

### Methods
- get_local_timestamp_oracle
- get_local_timestamp_oracle_mut
- get_local_read_ts
- get_local_write_ts
- get_and_step_local_write_ts
- peek_local_ts
- ensure_timeline_state
- remove_global_read_holds_storage
- remove_global_read_holds_compute
- ids_in_timeline
- validate_timeline
- get_timeline
- get_timelines
- timedomain_for

## timestamp_selection

### Description
Logic for selecting timestamps for various operations on collections.

### Structs
None

### Methods
- determine_timestamp
- least_valid_read
- least_valid_write
- largest_not_in_advance_of_upper

## util

### Description
Various utility methods used by the `Coordinator`. Ideally these are
all put in more meaningfully named modules.

### Structs
None

### Methods
- now
- now_datetime
- plan_statement
- declare
- describe
- verify_prepared_statement
- verify_portal
- verify_statement_revision
- clear_transaction
- allocate_transient_id

### Traits
- CoordTimestamp

File: src/adapter/src/coord/util.rs [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    ) -> TransactionStatus<mz_repr::Timestamp> {
+        let (drop_sinks, txn) = session.clear_transaction();
+        self.drop_sinks(drop_sinks).await;
+
+        // Release this transaction's compaction hold on collections.
+        if let Some(txn_reads) = self.txn_reads.remove(&session.conn_id()) {
+            self.release_read_hold(&txn_reads.read_holds).await;
+        }
+        txn
+    }
```
Comment (benesch) [inline src/adapter/src/coord/util.rs:166]: You mentioned this in the PR description, but I'd be +100 on moving these `mz_adapter::coord::message_procing::sql` module or somesuch. It's kind of funny to have such key methods in a module named `util`.

**Review by benesch — APPROVED** [review]
As before, didn't think about this too hard. I'm extremely 👍🏽 on getting this merged ASAP, and then we can iterate!

**Comment by frankmcsherry** [comment]
> As before, didn't think about this too hard. I'm extremely 👍🏽 on getting this merged ASAP, and then we can iterate!

Same for me, on both points!

File: src/adapter/src/coord/util.rs (hunk 1 above)
Comment (jkosh44) [reply src/adapter/src/coord/util.rs]: I took some stuff out and renamed it to `sql`.

**Comment by jkosh44** [comment]
Ok, once the build passes this is going to merge. Speak now or forever... open up a follow up PR.

**Pull request merged**